src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -24,7 +24,39 @@ internal static partial class OpenSsl
         private const string TlsCacheSizeCtxName = "System.Net.Security.TlsCacheSize";
         private const string TlsCacheSizeEnvironmentVariable = "DOTNET_SYSTEM_NET_SECURITY_TLSCACHESIZE";
         private const SslProtocols FakeAlpnSslProtocol = (SslProtocols)1;   // used to distinguish server sessions with ALPN
-        private static readonly ConcurrentDictionary<SslProtocols, SafeSslContextHandle> s_clientSslContexts = new ConcurrentDictionary<SslProtocols, SafeSslContextHandle>();
+        private static readonly ConcurrentDictionary<SslContextCacheKey, SafeSslContextHandle> s_clientSslContexts = new ConcurrentDictionary<SslContextCacheKey, SafeSslContextHandle>();
+
+        internal readonly struct SslContextCacheKey : IEquatable<SslContextCacheKey>
+        {
+            public readonly byte[]? CertificateThumbprint;
+            public readonly SslProtocols SslProtocols;
+
+            public SslContextCacheKey(SslProtocols sslProtocols, byte[]? certificateThumbprint)
+            {
+                SslProtocols = sslProtocols;
+                CertificateThumbprint = certificateThumbprint;
+            }
+
+            public override bool Equals(object? obj) => obj is SslContextCacheKey key && Equals(key);
+
+            public bool Equals(SslContextCacheKey other) =>
+                SslProtocols == other.SslProtocols &&
+                (CertificateThumbprint == null && other.CertificateThumbprint == null ||
+                 CertificateThumbprint != null && other.CertificateThumbprint != null && CertificateThumbprint.AsSpan().SequenceEqual(other.CertificateThumbprint));
+
+            public override int GetHashCode()
```

Comment:
What are the implications of a collision here, either by “luck” or a malicious client?

Reply:
Ah wait, I am assuming the client certificate has a valid chain at this point.

Reply:
The certificate is taken from SslStreamCertificateContext, so the (valid) chain should be already built by then.

In case of collision, we would attempt to reuse a session established with a different certificate. Note that the sessions are later cached per-host, so we would reuse sessions to the same host only.

Also note that this is the same mechanism we use in e.g. 

https://github.com/dotnet/runtime/blob/1815306c1b325a66189511f3ddecd45638c47681/src/libraries/System.Net.Security/src/System/Net/Security/SslSessionsCache.cs#L99-L107

Reply:
While the same certificate hash code is very unlikely (and we use it already) I'm wondering if that is good enough. Any thoughts on that @bartonjs ?

Reply:
Note that we moved to using SHA256 here https://github.com/dotnet/runtime/pull/102656#discussion_r1613804904, I think we should (in separate PR) do this for other places as well.

Reply:
Discussed offline, SHA256 is good enough.